### PR TITLE
Module7 task1

### PR DESCRIPTION
**Pull request by vamperor1**

### CODE REVIEW

File: project/src/components/app/app.tsx [hunk 1]
```diff
@@ -1,24 +1,31 @@
...
+import HistoryRouter from '../history-route/history-route';
+import browserHistory from '../../browser-history';
 
-type AppScreenProps = {
- offers: Offer[];
-}
+function App(): JSX.Element {
+  const {isOffersDataLoaded} = useAppSelector((state) => state);
+
+  if (!isOffersDataLoaded) {
```
Comment (Rodion-progs) [inline project/src/components/app/app.tsx:17]: В данном проекте может ок, так как почти все проекты содержат оферы, но если страниц больше, например, добавим страницу контактов, то зачем для страницы контактов показывать лоадрер пока грузятся офферы. У нас в проекте для каждой страницы своя страница лоадинга со своими условиями показа, а не для всего приложения.

File: project/src/components/login-screen/login-screen.tsx [hunk 2]
```diff
@@ -1,23 +1,82 @@
...
+  const onSubmit = (authData: AuthData) => {
+    dispatch(loginAction(authData));
+  };
+
+  const validatePassword = () => {
+    const passwordValue = passwordRef.current?.value;
+
+    if (passwordValue) {
+      if (!letters.test(passwordValue)) {
+        passwordRef.current?.setCustomValidity('Must contain at least one letter');
```
Comment (Rodion-progs) [inline project/src/components/login-screen/login-screen.tsx:28]: Проверку passwordRef.current можно вынести выше, чтобы внутри каждого if else не писать проверку current?.field. Именно тут некритично, но если у тебя был были бы проверки только с if то это было бы менее производительно. Т.е. тут замечание на будущее снова)
```
if (!letters.test(passwordValue)) {
        passwordRef.current?.setCustomValidity('Must contain at least one letter');
 } 
if (!numbers.test(passwordValue)) {
        passwordRef.current?.setCustomValidity('Must contain at least one number');
}
 if (spaces.test(passwordValue)) {
        passwordRef.current?.setCustomValidity('No spaces available');
}
if  {
        passwordRef.current?.setCustomValidity('');
}
```

Если было бы так, то это плохо было бы

Comment (Rodion-progs) [reply project/src/components/login-screen/login-screen.tsx:28]: А нет. Я все правильно написал. Надо вынести выге проверку всё же, так как у тебя два раза повторяется уже passwordRef.current? . Это не супер критично, но все же лучше поправить

Comment (Rodion-progs) [reply project/src/components/login-screen/login-screen.tsx:28]: А если passwordRef.current undefined, то ничего не должно происходить?

File: project/src/components/login-screen/login-screen.tsx [hunk 3]
```diff
@@ -1,23 +1,82 @@
...
+
   return (
     <div className="page page--gray page--login">
       <Header />
       <main className="page__main page__main--login">
         <div className="page__login-container container">
           <section className="login">
             <h1 className="login__title">Sign in</h1>
-            <form className="login__form form" action="#" method="post">
+            <form onSubmit={handleSubmit} className="login__form form" action="#" method="post">
```
Comment (Rodion-progs) [inline project/src/components/login-screen/login-screen.tsx:57]: Из- за метода POST могут возникнуть предостережения браузера о заполнености формы, когда будешь переходить по страницам кнопкой назад браузерной. У меня так было, только не в реакте. лучше get. Ну и насколько знаю method и action опциональные атрибуты формы

File: project/src/components/login-screen/login-screen.tsx [hunk 4]
```diff
@@ -1,23 +1,82 @@
...
               <div className="login__input-wrapper form__input-wrapper">
                 <label className="visually-hidden">E-mail</label>
```
Comment (Rodion-progs) [inline project/src/components/login-screen/login-screen.tsx:59]: Про связывание input и label не стоит забывать, нативное поведение браузера которое делает сайт более доступным. input и label можно связать id и for либо вкладывать input в label

File: project/src/components/login-screen/login-screen.tsx [hunk 5]
```diff
@@ -1,23 +1,82 @@
...
+                  name="email"
+                  placeholder="Email"
+                  required
+                />
               </div>
               <div className="login__input-wrapper form__input-wrapper">
                 <label className="visually-hidden">Password</label>
-                <input className="login__input form__input" type="password" name="password" placeholder="Password" required />
+                <input
+                  onChange={validatePassword}
```
Comment (Rodion-progs) [inline project/src/components/login-screen/login-screen.tsx:72]: validatePassword вызывает каждый раз на onChange, тога зачем нужен passwordRef, если можно сделать evt.target.setCustomValidity('Must contain at least one number'); 
или
this.setCustomValidity('Must contain at least one number');

File: project/src/components/offer-screen/offer-screen.tsx [hunk 6]
```diff
@@ -1,37 +1,65 @@
...
+  const {isNearOffersDataLoaded} = useAppSelector((state) => state);
+  const {isReviewsDataLoaded} = useAppSelector((state) => state);
+
+  useEffect (() => {
+    dispatch(fetchOfferDetailsAction(id));
+    dispatch(fetchNearOffersAction(id));
+    dispatch(fetchReviewsAction(id));
+  }, [dispatch, id]);
+
+  console.log(isOfferDetailsDataLoaded, isNearOffersDataLoaded, isReviewsDataLoaded);
```
Comment (Rodion-progs) [inline project/src/components/offer-screen/offer-screen.tsx:36]: Не забывай убирать логи)

File: project/src/components/offer-screen/offer-screen.tsx [hunk 7]
```diff
@@ -1,37 +1,65 @@
...
   const {name, isPro, avatarUrl} = offer.host;
+
+  const isAuthorized = authStatus === AuthorizationStatus.Auth;
+
   const avatarClass = `property__avatar-wrapper ${
     isPro ? 'property__avatar-wrapper--pro' : null
   } user__avatar-wrapper`;
+
   const shownType = type.slice(0, 1).toUpperCase() + type.slice(1);
   const shownImages = images.length > 6 ? images.slice(0, 6) : images;
```
Comment (Rodion-progs) [inline project/src/components/offer-screen/offer-screen.tsx:61]: 6 вынести в константы.

File: project/src/services/error-handle.ts [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+  if (response) {
+    switch (response.status) {
+      case HTTP_CODE.BAD_REQUEST:
+        toast.info(response.data.error);
+        break;
+      case HTTP_CODE.UNAUTHORIZED:
+        // TODO: Не знаю, какую обработку этой ошибки можно здесь сделать.
+        // Всё реализуется на уровне функционала компонентов.
+
+        // toast.info(response.data.error);
```
Comment (Rodion-progs) [inline project/src/services/error-handle.ts:24]: toast.info и оставить. Если что пойдёт не так, юзер будет знать и ничего не сломается по логике от уведомления всплывшего

File: project/src/store/api-actions.ts [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+  'user/login',
+  async ({id, comment, rating, resetForm, unblockForm}, {dispatch, extra: api}) => {
+    try {
+      const {data} = await api.post<Review[]>(`${APIRoute.Comments}/${id}`, {comment, rating});
+      dispatch(loadReviews(data));
+      resetForm();
+      unblockForm();
+    } catch(error) {
+      errorHandle(error);
+      unblockForm();
```
Comment (Rodion-progs) [inline project/src/store/api-actions.ts:96]: unblockForm(); можно в блок finaly перенести

File: package.json [hunk 10]
```diff
@@ -0,0 +1 @@
+{}
```
Comment (Rodion-progs) [inline package.json:1]: Удалить этот файл нужно и package-lock рядом тоже

File: project/src/components/offer-reviews-form/offer-reviews-form.tsx [hunk 11]
```diff
@@ -1,30 +1,93 @@
...
+}
+
+function OfferReviewsForm({ id }: OfferReviewsFormProps): JSX.Element {
+  const [rating, setRating] = useState('');
   const [review, setReview] = useState('');
+  const dispatch = useAppDispatch();
+
+  const formRef = useRef<HTMLFormElement | null>(null);
+  const fieldsetRef = useRef<HTMLFieldSetElement | null>(null);
+  const isReviewValid = rating.length > 0 && (review.length > 50 && review.length < 300);
```
Comment (Rodion-progs) [inline project/src/components/offer-reviews-form/offer-reviews-form.tsx:19]: 50 и 300 в константы

**Review by Rodion-progs — APPROVED** [review]

File: project/src/components/login-screen/login-screen.tsx (hunk 2 above)
Comment (vamperor1) [reply project/src/components/login-screen/login-screen.tsx:28]: Ты имеешь в виду вот так?
 if (passwordValue **&& passwordRef.current**) {
      if (!letters.test(passwordValue)) {
        passwordRef.current.setCustomValidity('Must contain at least one letter');
      } else if (!numbers.test(passwordValue)) {
        passwordRef.current.setCustomValidity('Must contain at least one number');
      } else if (spaces.test(passwordValue)) {
        passwordRef.current.setCustomValidity('No spaces available');
      } else {
        passwordRef.current.setCustomValidity('');
      }
    }
  };

Comment (vamperor1) [reply project/src/components/login-screen/login-screen.tsx:28]: > А если passwordRef.current undefined, то ничего не должно происходить?

да вроде нет, если он undefined, то значит поле пароля не отрисовалось и вводить туда ничего не получится )

File: project/src/components/login-screen/login-screen.tsx (hunk 3 above)
Comment (vamperor1) [reply project/src/components/login-screen/login-screen.tsx:57]: Ну экшен по авторизации тут полностью копи-паст с демо-проекта.

А разве в get можно отправить авторизационные данные?
axios.get(url[, config]);
axios.post(url[, **data**[, config]]);

File: project/src/components/login-screen/login-screen.tsx (hunk 4 above)
Comment (vamperor1) [reply project/src/components/login-screen/login-screen.tsx:59]: Ну верстка нам досталась в готовом виде.
Видимо, авторы курса решили, что здесь это не столь важно.

File: project/src/components/login-screen/login-screen.tsx (hunk 5 above)
Comment (vamperor1) [reply project/src/components/login-screen/login-screen.tsx:72]: Ну он используется для передачи пароля на сервер.
Раз он уже есть, почему его не использовать и для валидации? )

File: package.json (hunk 10 above)
Comment (vamperor1) [reply package.json:1]: Почему то еще на первом занятии сказали, что это не надо трогать... Мы с тобой этого не поняли, но раз сказали, то решили оставить, он не мешает )

Comment (Rodion-progs) [reply package.json:1]: Так а почему он в этом ПР появился снова, как измененный?

File: project/src/components/login-screen/login-screen.tsx (hunk 2 above)
Comment (Rodion-progs) [reply project/src/components/login-screen/login-screen.tsx:28]: Можно и так. Я бы так оставил 
```

 const validatePassword = () => {
    if (!passwordRef.current) {
       return;
    }

    const passwordValue = passwordRef.current.value;

    if (passwordValue) {
      if (!letters.test(passwordValue)) {
        passwordRef.current.setCustomValidity('Must contain at least one letter');
      } else if () {
        ...
      }
   }
}
```

File: project/src/components/login-screen/login-screen.tsx (hunk 3 above)
Comment (Rodion-progs) [reply project/src/components/login-screen/login-screen.tsx:57]: Я имел в виду, что атрибуты формы тут неважны, так как ты все равно будешь отправлять через ajax. Авторизационные данные естественно axios.post. В самой форме method="post" можно было не указывать, хотя может и стоит, так как дополнительная проверка на безопасность будет со стороны браузера. На странице логина думаю эта проверка не супер важная.

File: project/src/components/login-screen/login-screen.tsx (hunk 5 above)
Comment (Rodion-progs) [reply project/src/components/login-screen/login-screen.tsx:72]: До инпута можно добрать через e.target.password.value и e.target.email.value

**Review by Rodion-progs — APPROVED** [review]
